src/Network/NWFramerMessage.cs

```diff
@@ -0,0 +1,137 @@
+//
+// NWFramer.cs: Bindings the Network nw_framer_t API.
+//
+// Authors:
+//   Manuel de la Pena (mandel@microsoft.com)
+//
+// Copyright 2019 Microsoft
+//
+using System;
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+using ObjCRuntime;
+using Foundation;
+using CoreFoundation;
+
+using OS_nw_framer=System.IntPtr;
+using OS_nw_protocol_metadata=System.IntPtr;
+using OS_dispatch_data=System.IntPtr;
+using OS_nw_protocol_definition=System.IntPtr;
+using OS_nw_protocol_options=System.IntPtr;
+using OS_nw_endpoint=System.IntPtr;
+using OS_nw_parameters=System.IntPtr;
+
+namespace Network {
+
+	[TV (13,0), Mac (10,15), iOS (13,0), Watch (6,0)]
+	public class NWFramerMessage : NWProtocolMetadata {
+		internal NWFramerMessage (IntPtr handle, bool owns) : base (handle, owns) {}
+
+		[DllImport (Constants.NetworkLibrary)]
+		static extern OS_nw_protocol_metadata nw_framer_protocol_create_message (OS_nw_protocol_definition definition);
+
+		public static NWFramerMessage CreateMessage (NWProtocolDefinition protocolDefinition)
+		{
+			if (protocolDefinition == null)
+				throw new ArgumentNullException (nameof (protocolDefinition));
+			return new NWFramerMessage (nw_framer_protocol_create_message (protocolDefinition.Handle), owns: true);
+		}
```

Comment:
wrong place, sorry again :)

I think `nw_framer_protocol_create_message` should be `NWFramerProtocol.CreateMessage`

Reply:
So this, is a complicated and API because it is not very consistent with the naming. For what I have been able to see, we have:

* nw_framer_protocol_create_message which is the only C call that has the nw_framer_protocol_* pattern. And that takes a OS_nw_protocol_definition which is static.
* nw_framer_ which is the pattern used by the C library when working with OS_nw_framer which is what I'm mapping to NWFramer.
* nw_framer_message_* pattern. Which maps got a NWFramerMessage which is a NWProtocolMetadata.

Looking at swift, the code is not very clear because it does a lot of magic, and they expose a NWFramerProtocol convenience class, that contains a NWFramer and hooks to all the different callbacks.  My reasoning between moving the CreateMessage **static** method to the NWFramer is because in order to work with the NWFramer, you need to be able to write NWFramerMessage (which is just a specific type of metadata) which is later passed to the NWConneciton, something like the following:

```csharp
var message = protocol.CreateMessage (GameMessageType.Move);
var context = new NWContentContext ("Move");
context.SetMetadata (message);

connection.Send (Encoding.UTF8.GetBytes (move), context, true, (e) => {
	if (e != null) {
		Console.WriteLine ($"Error sending move: error number {e.ErrorCode} domain {e.ErrorDomain}");
	}
});
```

In that case we are creating a NWFramerMessage that has a header which contains the GamerMessageType (it is added as a string in one of the headers using a key).

I would consider nw_framer_protocol_ to be similar to nw_framer_ since the naming of the API is inconsistent. Now, where to place the CreateMessage is mostly on what we want to expose as an API. In the  swift code there is no public implementation of the NWFramer, but if you look at it using Xcode you get the [following](https://gist.github.com/mandel-macaque/abda55c3b12ec50a9e5d2766a7274b0b).

There are using two types, NWFramer, and a wrapper NWProtocolFramer that contains and instance of the NWFramer and can be init with a NWProtocolDefinition. I don't want to write magic, at least not at this stage and not with the same pattern they have used (they have inner classes, so rather than a NWTlsProtocolOptions, they have an Options class inside the NWTlsProtocol). I have no preference per se, either one is good as long as we do not create NWFramerProtocol, but use NWFramer. I think that NWFramerMessage.Create (definition) is nice, even when you can create the message in two diff classes.

Reply:
ok :) just so people reading the source does not wonder why (and try to reach the same conclusion as you did) please add a short code comment, which can be a link back to your detailed answer here